core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -699,19 +701,28 @@ private[kafka] class Processor(val id: Int,
   private def processCompletedSends() {
     selector.completedSends.asScala.foreach { send =>
       try {
-        val resp = inflightResponses.remove(send.destination).getOrElse {
+        val response = inflightResponses.remove(send.destination).getOrElse {
           throw new IllegalStateException(s"Send for ${send.destination} completed, but not in `inflightResponses`")
         }
-        updateRequestMetrics(resp)
+        updateRequestMetrics(response)
+
+        // Invoke send completion callback
+        response match {
```

Comment:
Similar to the comment above. If you find yourself matching on types, it is worth considering whether we can instead add a method to the type. Here we could have `Response.onComplete`, and the default implementation can do nothing.